### PR TITLE
Make k8s meta processor initialisation asynchronous

### PR DESCRIPTION
**Pull request by ChrsMark**
## What does this PR do?
This PR changes `add_kubernetes_metadata` processor so as to get initialised asynchronously and being able to handle `node not ready` situations.

## Why is it important?
This will be solve the problem of having to restart Beat in case the processor was not initialised if the node was not ready yet when Beat started. 

## How to test this PR locally
With the use of Minikube:
1. `minikube start`
2. Have a `minikube ssh` in one tab so as to handle k8s health up and down
3. Build Filebeat on the host where Minikube runs
4. Use the kubeconfig of Minikube: `export KUBECONFIG=~/.kube/config`
5. Make sure that `add_kubernetes_metadata` processor is enabled:
```
processors:
  - add_kubernetes_metadata: ~
```
6.  Start filebeat and expect to see `2020-02-18T12:25:12.825+0200	INFO	add_kubernetes_metadata/kubernetes.go:77	add_kubernetes_metadata: kubernetes env detected, with version: v1.17.0` in your logs, meaning that processor was able to reach k8s API.
7. Stop filebeat
8. In the session you have inside Minikube vm run `sudo systemctl stop docker` to bring docker down and make k8s API unreachable
9. Start filebeat again and you should see (after ~30secs) sth like `2020-02-18T12:27:16.958+0200	INFO	add_kubernetes_metadata/kubernetes.go:89	add_kubernetes_metadata: could not detect kubernetes env: Get https://192.168.64.6:8443/version?timeout=32s: dial tcp 192.168.64.6:8443: connect: connection refused`
10. Stop filebeat and start it again.
11. Wait 3-4 seconds and go to Minkube vm and start docker daemon again: `sudo systemctl start` docker.
12. In Filebeat's logs you should expect to see that the processor reached k8s API like in step 6.

## Related issues

- Closes https://github.com/elastic/beats/issues/16064

### CODE REVIEW

File: libbeat/processors/add_kubernetes_metadata/kubernetes.go [hunk 1]
```diff
@@ -108,71 +111,80 @@ func New(cfg *common.Config) (processors.Processor, error) {
...
+			},
+			DeleteFunc: func(obj interface{}) {
+				pod := obj.(*kubernetes.Pod)
+				log.Debugf("Removing pod: %s/%s", pod.GetNamespace(), pod.GetName())
+				processor.removePod(pod)
+			},
+		})
+
+		if err := watcher.Start(); err != nil {
+			return
```
Comment (jsoriano) [inline libbeat/processors/add_kubernetes_metadata/kubernetes.go]: Log this error at least at the debug level?

File: libbeat/processors/add_kubernetes_metadata/kubernetes.go [hunk 2]
```diff
@@ -108,71 +111,80 @@ func New(cfg *common.Config) (processors.Processor, error) {
...
-	if err != nil {
-		if kubernetes.IsInCluster(config.KubeConfig) {
-			log.Debugf("Could not create kubernetes client using in_cluster config: %+v", err)
-		} else if config.KubeConfig == "" {
-			log.Debugf("Could not create kubernetes client using config: %v: %+v", os.Getenv("KUBECONFIG"), err)
-		} else {
-			log.Debugf("Could not create kubernetes client using config: %v: %+v", config.KubeConfig, err)
+	// complete processor's initialisation asynchronously so as to re-try on failing k8s client initialisations in case
+	// the k8s node is not yet ready.
+	go func(processor *kubernetesAnnotator) {
```
Comment (jsoriano) [inline libbeat/processors/add_kubernetes_metadata/kubernetes.go]: In other processors this code lives in a [`processor.init()` function](https://github.com/elastic/beats/blob/c5a568f761b63414f3415e8f6eceb13b84919dbf/libbeat/processors/add_cloud_metadata/add_cloud_metadata.go#L76), we might do the same here.
```suggestion
	go processor.init(config)
```

File: libbeat/processors/add_kubernetes_metadata/kubernetes.go [hunk 3]
```diff
@@ -108,71 +111,80 @@ func New(cfg *common.Config) (processors.Processor, error) {
...
-		return processor, nil
-	}
 
-	if !isKubernetesAvailable(client) {
-		return processor, nil
-	}
-
-	matchers := NewMatchers(config.Matchers)
+		connectionAttempts := 1
+		for ok := true; ok; ok = !isKubernetesAvailable(client) {
```
Comment (jsoriano) [inline libbeat/processors/add_kubernetes_metadata/kubernetes.go]: `isKubernetesAvailable` is logging the result on every attempt, if kubernetes is not available it is going to log it 10 times at the info level. Should we only log it on the last attempt or something like this?

File: libbeat/processors/add_kubernetes_metadata/kubernetes.go (hunk 2 above)
Comment (ChrsMark) [reply libbeat/processors/add_kubernetes_metadata/kubernetes.go]: Thanks for pointing to this!

**Comment by ChrsMark** [comment]
Changed some stuff and added testing notes. Tested it with minikube with the process described in the PR's description.

File: libbeat/processors/add_kubernetes_metadata/kubernetes.go [hunk 4]
```diff
@@ -37,9 +40,13 @@ import (
 )
 
 const (
-	timeout = time.Second * 5
+	timeout                = time.Second * 5
+	selector               = "kubernetes"
+	checkNodeReadyAttempts = 10
 )
 
+var once sync.Once
```
Comment (jsoriano) [inline libbeat/processors/add_kubernetes_metadata/kubernetes.go]: Don't use a global for this please, `add_kubernetes_metadata` could be defined in several parts of the config (or in some dynamic config), and only the first appearance would be initialized.

File: libbeat/processors/add_kubernetes_metadata/kubernetes.go [hunk 5]
```diff
@@ -22,8 +22,11 @@ package add_kubernetes_metadata
 import (
 	"fmt"
 	"os"
+	"sync"
 	"time"
 
+	"github.com/cloudflare/cfssl/log"
+
 	"github.com/elastic/beats/libbeat/common/kubernetes/metadata"
```
Comment (jsoriano) [inline libbeat/processors/add_kubernetes_metadata/kubernetes.go]: Nit. Imports order, this libbeat import should be below with the other ones.

**Comment by ChrsMark** [comment]
While fixing the import order I noticed that one import("github.com/cloudflare/cfssl/log") was not used 🤔 so I removed it and also tried to unify the logging using the processor's logger along. Sorry for the back-and-forth.

**Comment by jsoriano** [comment]
> While fixing the import order I noticed that one import("github.com/cloudflare/cfssl/log") was not used thinking so I removed it and also tried to unify the logging using the processor's logger along. Sorry for the back-and-forth slightly_smiling_face .

Good catch!

**Pull request merged**